### PR TITLE
[24.1] Document 1:N fanout transforms

### PR DESCRIPTION
**Pull request by kbatuigas**
- Additions to [Golang](https://deploy-preview-374--redpanda-docs-preview.netlify.app/24.1/reference/data-transform-golang-sdk/#totopic) and [Rust](https://deploy-preview-374--redpanda-docs-preview.netlify.app/24.1/reference/data-transform-rust-sdk/#recordwriter) SDK reference pages
- Update [How Transforms Work](https://deploy-preview-374--redpanda-docs-preview.netlify.app/24.1/develop/data-transforms/how-transforms-work/)
- Update [Run Transforms](https://deploy-preview-374--redpanda-docs-preview.netlify.app/24.1/develop/data-transforms/run-transforms/) tutorial

Resolves https://github.com/redpanda-data/documentation-private/issues/2234

### CODE REVIEW

File: modules/develop/partials/run-transforms.adoc [hunk 1]
```diff
@@ -443,7 +445,7 @@ See xref:reference:public-metrics-reference.adoc[]
 
 - Transforms have no external access to disk or network resources.
 - Only single record transforms is supported, but multiple output records from a single input record is supported. For aggregations, joins, or complex transformations, use Apache Flink.
-- Only a single output topic is supported.
+- Up to 16 output topics is supported.
```
Comment (Deflaimun) [inline modules/develop/partials/run-transforms.adoc]:
```suggestion
- Up to 8 output topics are supported.
```

File: modules/reference/pages/data-transform-rust-sdk.adoc [hunk 2]
```diff
@@ -191,8 +191,15 @@ impl<'a> RecordWriter<'a> {
     // Creates a new [`RecordWriter`] using the specified `sink`.
     pub fn new(sink: &'a mut dyn RecordSink) -> Self;
 
-    /// Write a record to the output topic returning any errors.
+    /// Write a record to the default output topic, returning any errors.
```
Comment (Deflaimun) [inline modules/reference/pages/data-transform-rust-sdk.adoc:194]: was this the one that they provided which topic it was?

File: modules/develop/partials/run-transforms.adoc (hunk 1 above)
Comment (rockwotj) [reply modules/develop/partials/run-transforms.adoc]: Actually this is 8, there is a difference than the RFC and was lowered (it's kind of an arbitrary limit)

File: modules/reference/pages/data-transform-golang-sdk.adoc [hunk 3]
```diff
@@ -93,6 +93,14 @@ type WriteEvent interface {
 
 The `WriteEvent` type contains information about a record that was written.
 
+==== WriteOpt
+
+```go
+type WriteOpt = func(c *writeOpts)
+```
+
+The `WriteOpt` type provides options to customize a write, for example, specify an output topic to write to.
```
Comment (rockwotj) [inline modules/reference/pages/data-transform-golang-sdk.adoc:111]: We also need to document the `ToTopic` function, which is of type `WriteOpt`

File: modules/develop/partials/run-transforms.adoc [hunk 4]
```diff
@@ -246,7 +246,7 @@ func filterValidJson(e transform.WriteEvent, w transform.RecordWriter) error {
 	if json.Valid(e.Record().Value) {
 		return w.Write(e.Record())
 	}
-	return nil
+	return w.Write(e.Record(), transform.ToTopic("invalid-json"))
```
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc:250]: Here and for Rust, should we add a comment saying:

```
// Send invalid records to separate topic
```

File: modules/reference/pages/data-transform-rust-sdk.adoc (hunk 2 above)
Comment (kbatuigas) [reply modules/reference/pages/data-transform-rust-sdk.adoc:194]: Not sure, @rockwotj can you comment?

File: modules/develop/partials/run-transforms.adoc [hunk 5]
```diff
@@ -354,6 +355,9 @@ fn filter_valid_json(event: WriteEvent, writer: &mut RecordWriter) -> Result<()>
     let value = event.record.value().unwrap_or_default();
     if serde_json::from_slice::<serde_json::Value>(value).is_ok() {
         writer.write(event.record)?;
+    } else {
+		// Send invalid records to separate topic
```
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc]: Horizontal alignment

File: modules/reference/pages/data-transform-rust-sdk.adoc (hunk 2 above)
Comment (rockwotj) [reply modules/reference/pages/data-transform-rust-sdk.adoc:194]: The "default" output topic is the 0th output topic in their config file or in their cli deploy `rpk transform deploy --output-topic foo --output-topic bar` The default topic in this case is `foo`.

**Review by rockwotj — APPROVED** [review]

File: modules/develop/pages/data-transforms/how-transforms-work.adoc [hunk 6]
```diff
@@ -14,7 +14,7 @@ Data transforms let you run common data streaming tasks, like filtering, scrubbi
 
 Data transforms use https://webassembly.org/[WebAssembly^] (Wasm) engines inside a Redpanda broker, allowing Redpanda to control the entire transform lifecycle. For example, Redpanda can stop and start transforms when partitions are moved or to free up system resources for other tasks.
 
-Data transforms take data from an input topic and map it to an output topic. For each topic partition, a leader is responsible for handling the data. Redpanda runs a virtual machine (VM) on the same CPU core (shard) as these partition leaders to execute the transforms function.
+Data transforms take data from an input topic and map it to multiple output topics. For each topic partition, a leader is responsible for handling the data. Redpanda runs a virtual machine (VM) on the same CPU core (shard) as these partition leaders to execute the transforms function.
```
Comment (mattschumpert) [inline modules/develop/pages/data-transforms/how-transforms-work.adoc]: @kbatuigas  

"one ore more output topics".  It's not always multiple.

"Web Assembly Virtual Machine" to be specific

File: modules/reference/pages/data-transform-golang-sdk.adoc [hunk 7]
```diff
@@ -20,6 +21,14 @@ The `OnRecordWritten` function registers a callback of type <<onrecordwrittencal
 
 The function should be called in a package's `main` function to register the transform function that will be applied.
 
+==== ToTopic
+
+```go
+func ToTopic(topic string) WriteOpt
```
Comment (mattschumpert) [inline modules/reference/pages/data-transform-golang-sdk.adoc:27]: Sidenote: I think this is a change to the API (@rockwotj ?) which should be documented in 'incompatible changes' in 24.1 (even though WASM was in beta in 23.3 we should still note this breaking change).

**Review by mattschumpert — APPROVED** [review]

File: modules/reference/pages/data-transform-golang-sdk.adoc (hunk 7 above)
Comment (rockwotj) [reply modules/reference/pages/data-transform-golang-sdk.adoc:27]: It's not incompatible, it's purely additive, but you will have to upgrade the SDK

Comment (kbatuigas) [reply modules/reference/pages/data-transform-golang-sdk.adoc:27]: @rockwotj do we need to provide guidance on upgrading? In the [prereqs](https://deploy-preview-374--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/#prerequisites) we say "A Redpanda cluster running version 23.3," could we change that to "23.3 and later"?

Comment (rockwotj) [reply modules/reference/pages/data-transform-golang-sdk.adoc:27]: I think there is an open question of if we should be versioning the SDK docs or not, but these APIs will only be able to be used in 24.1

**Review by Feediver1 — APPROVED** [review]
Looks like some updates suggested by @mattschumpert remain to be implemented. Otherwise, LGTM.

**Review by Deflaimun — APPROVED** [review]

**Comment by mattschumpert** [comment]
@kbatuigas @rockwotj weren;t we going to update the tutorial to show an example using multiple output topics?  I think that would be helpful.

**Comment by rockwotj** [comment]
Yes we did - the validation example uses multiple output topics

**Pull request merged**